# Docker Compose

![Status](https://img.shields.io/static/v1.svg?label=Status&message=Finished&color=brightgreen)
[![Source](https://img.shields.io/static/v1.svg?label=GitHub&message=Source&color=181717&logo=GitHub)](https://github.com/particle1331/ok-transformer/blob/master/docs/nb/dk/01-compose.ipynb)
[![Stars](https://img.shields.io/github/stars/particle1331/ok-transformer?style=social)](https://github.com/particle1331/ok-transformer)

---

## Introduction

Multi-container application requires setting up run and builds, volumes, as well as networking between multiple services. This can be tedious to setup and teardown for each build and run, e.g. during development when refactoring related interfaces between services. 
[Docker Compose](https://docs.docker.com/compose/) allows us to collect all of this in an intuitive and human-readable format using a `docker-compose.yaml` file. Docker Compose also automatically takes care of networking between containers as well as logs and status for the whole ensemble. The idea is that multiple containers which function as a single entity should be managed as a single entity. To demonstrate a use case, we create two services consisting of a simple web application and a [redis](https://redis.io/) server as in-memory database.

## FastAPI app

First we create a simple FastAPI web app. This installs pipenv on system python which is generally not recommended, but is fine here because of container isolation. This will demonstrate how to run a web application using Docker. This app defines a single endpoint: 

```python
@app.get("/")
def hello():
    return {"message": "hello, world!"}
```

### Build

Setting `/usr/app` as the **working directory**. This is where all subsequent build commands will be executed. The copy command can be confusing. See the following figure. Note that we structure the application so that source files are separated in a `/src` folder while the other project files are in the root folder.

In [78]:
!cat simple-web/Dockerfile

FROM python:3.9.15-slim

RUN pip install -U pip
RUN pip install pipenv

WORKDIR /usr/app

COPY Pipfile Pipfile.lock ./
RUN pipenv install --system --deploy

COPY ./src/ ./src/

CMD uvicorn src.main:app --port 8080 --host 0.0.0.0


```{figure} diagrams/01-copy.svg
---
width: 330px
name: copy
---
Copies files from a local path to a path inside the container. The trailing `/` is important.
```

Building:

In [90]:
!docker build simple-web -t okt/simple-web --quiet
!docker run --rm -d okt/simple-web:latest

sha256:b24449de78e7f20e0956c0fcec900608e09778c01d00c03b6f13a5eb22a1e447
fc22c2dee362fd29039db28f0521aa7691cb6459ceade203f00a3c050b10775f


Checking if the file structure is as intended:

In [88]:
!tree simple-web

simple-web
├── Dockerfile
├── Pipfile
├── Pipfile.lock
└── src
    └── main.py

2 directories, 4 files


In [91]:
!docker exec fc22c2dee362 apt update > /dev/null
!docker exec fc22c2dee362 apt install tree > /dev/null
!docker exec fc22c2dee362 tree

.
├── Pipfile
├── Pipfile.lock
└── src
    ├── __pycache__
    │   └── main.cpython-39.pyc
    └── main.py

2 directories, 4 files


### Setting up ports

In [92]:
!docker logs fc22c2dee362

INFO:     Started server process [7]
INFO:     Waiting for application startup.
INFO:     Application startup complete.
INFO:     Uvicorn running on http://0.0.0.0:8080 (Press CTRL+C to quit)


Note that this port is not accessible to us. To access this, we have to publish ports:

In [93]:
!http :8080/


http: LogLevel.ERROR: ConnectionError: HTTPConnectionPool(host='localhost', port=8080): Max retries exceeded with url: / (Caused by NewConnectionError('<urllib3.connection.HTTPConnection object at 0x1060cdbd0>: Failed to establish a new connection: [Errno 61] Connection refused')) while doing a GET request to URL: http://localhost:8080/




<br>

```{figure} diagrams/01-port-mapping.svg
---
width: 800px
name: port
---
Route incoming requests to port 3000 on local host to port 8080 inside the container.
```

In [98]:
!docker run --rm -d -p 3000:8080 okt/simple-web:latest

9cc1e3571ce17a64ec55cc24dfaa22f27280ed85e8fd519aaf7c90550227b5ee


In [100]:
!http GET :3000/

HTTP/1.1 200 OK
content-length: 27
content-type: application/json
date: Fri, 24 Mar 2023 16:42:58 GMT
server: uvicorn

{
    "message": "hello, world!"
}




In [101]:
!docker stop 9cc1e3571

9cc1e3571


## Multiple services

Here we will build a web server which extends a bit the above simple web app to display the number of visits of a page. This responds to HTTP requests and generates HTML to show inside a browser. To actually store the number of times a page is visited, we will make use of a little redis server. Note that we can install the redis server inside the web server itself. But this setup does not scale well, e.g. to having multiple servers that depend on the same number of visits value. Also, we lose all our data when the web server fails. So this is a more robust setup.

### Web server

Our web server will implement the following endpoint. Notice that the host is `redis-server` instead of a URL. This is one magic of using `docker-compose` where we only need to specify the name of our service. This is discussed in the next section.

```python
app = FastAPI()
r = redis.Redis(host='redis-server', port=6379)
if not r.exists('visits'):
    r.set('visits', 0)


@app.get("/", response_class=HTMLResponse)
def home():
    visits = int(r.get('visits')) + 1
    r.set('visits', visits)
    return f"Number of visits: {visits}"
```

The Dockerfile for this service is almost the same as our simple web app above. Note that we deploy using [gunicorn with uvicorn workers](https://fastapi.tiangolo.com/deployment/server-workers/) as recommended in the docs. 

In [165]:
!cat visits/app/Dockerfile

FROM python:3.9.15-slim

RUN pip install -U pip
RUN pip install pipenv

WORKDIR /usr/app

COPY Pipfile Pipfile.lock ./
RUN pipenv install --system --deploy

COPY ./src/ ./src/

CMD gunicorn src.main:app --workers 1 --worker-class uvicorn.workers.UvicornWorker --bind 0.0.0.0:8081


### Redis DB

For our redis server, we simply pull the image from Docker Hub.

In [102]:
!docker run --rm -d redis

10035603adf079b136d42c013e8b86fb7037806aa2f2f93945efec0263b4e939


In [103]:
!docker logs 10035603ad

1:C 24 Mar 2023 19:08:17.571 # oO0OoO0OoO0Oo Redis is starting oO0OoO0OoO0Oo
1:C 24 Mar 2023 19:08:17.572 # Redis version=7.0.10, bits=64, commit=00000000, modified=0, pid=1, just started
1:C 24 Mar 2023 19:08:17.572 # Warning: no config file specified, using the default config. In order to specify a config file use redis-server /path/to/redis.conf
1:M 24 Mar 2023 19:08:17.572 * monotonic clock: POSIX clock_gettime
1:M 24 Mar 2023 19:08:17.573 * Running mode=standalone, port=6379.
1:M 24 Mar 2023 19:08:17.573 # Server initialized
1:M 24 Mar 2023 19:08:17.576 * Ready to accept connections


Setting visit count initially to zero:

In [105]:
!docker exec 10035603ad redis-cli set visits 0

OK


In [106]:
!docker exec 10035603ad redis-cli get visits

0


## Docker Compose

As mentioned, to manage our two services easily, we use `docker-compose` version 3. Here services means our two containers which perform two specific purposes. To configure, we have to specify a `docker-compose.yml` file. 

In [166]:
!cat visits/docker-compose.yml

version: '3'
services:
  redis-server:
    image: redis
  web-server:
    restart: on-failure
    build: app
    ports:
      - 3001:8081


As mentioned above, the redis server built from the `redis` image (pulled if not locally available) will be hosted on the host `'redis-server'` which will be accessible from the `web-server` app. The build path `app/` specifies the location of the Dockerfile for this service relative to the YAML file. Finally, we have to publish the container port 8081 to 3001 on our local machine which runs the Docker server. The ports is a list as indicated by `-`, so we can publish as many ports as we want.

In [167]:
!tree visits

visits
├── app
│   ├── Dockerfile
│   ├── Pipfile
│   ├── Pipfile.lock
│   └── src
│       └── main.py
└── docker-compose.yml

3 directories, 5 files


If we have another service that we manually build (i.e. not simply pull), it will be in the same level as `app/`. The `docker-compose.yml` file is on the same level as the service directories. Here we only have one.

### Compose up

The `--build` flag rebuilds the images. This is not really necessary for a first build since the images are automatically built with `docker-compose up` if the images are not available. Running the entire ensemble:

In [181]:
!docker-compose --project-directory visits up -d --build

[+] Building 0.0s (0/0)                                                         
[+] Building 0.0s (0/1)                                                         
[+] Building 0.2s (2/3)                                                         
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 32B                                        0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B                                            0.0s
 => [internal] load metadata for docker.io/library/python:3.9.15-slim      0.1s
[+] Building 0.3s (2/3)                                                         
 => [internal] load build definition from Dockerfile                       0.0s
 => => transferring dockerfile: 32B                                        0.0s
 => [internal] load .dockerignore                                          0.0s
 => => transferring context: 2B     

It is important to note the other services which has no change will not be rebuilt. These services can persist data and state while rebuilt services have reset state. To view running processes, we use:

In [182]:
!docker-compose --project-directory visits ps

NAME                    COMMAND                  SERVICE             STATUS              PORTS
visits-redis-server-1   "docker-entrypoint.s…"   redis-server        running             6379/tcp
visits-web-server-1     "/bin/sh -c 'gunicor…"   web-server          running             0.0.0.0:3001->8081/tcp


Due to container isolation, a fresh redis DB is instantiated when the services are up:

In [183]:
!http :3001

HTTP/1.1 200 OK
content-length: 19
content-type: text/html; charset=utf-8
date: Fri, 24 Mar 2023 20:15:32 GMT
server: uvicorn

Number of visits: 1




In [184]:
!http :3001

HTTP/1.1 200 OK
content-length: 19
content-type: text/html; charset=utf-8
date: Fri, 24 Mar 2023 20:15:32 GMT
server: uvicorn

Number of visits: 2




<br>

```{figure} diagrams/01-containers.png
---
name: containers
---
Running composed services from Docker Desktop.
```


```{figure} diagrams/01-compose-monitoring.png
---
name: compose-monitoring
---
Monitoring running services. Note that a terminal can also be opened from inside the container.
```

### Restart policies

Note that a web service typically crashes for a variety of reasons. Oftentimes this can happen rarely so that we want to restart one particular service immediately after it crashes. In the `docker-compose.yml` we specified `on-failure` as the restart policy. This means the service is restarted every time it fails with nonzero exit status, this is different from being stopped which has exit status `0`.

<br>

```{figure} diagrams/01-restart-policies.png
---
width: 650px
name: restart-policies
---
[Restart policies](https://docs.docker.com/config/containers/start-containers-automatically/#use-a-restart-policy) for containers in Docker.
```

To demonstrate, we defined an endpoint which results in a failure as described:

```python
@app.get("/kill")
async def kill_uvicorn():
    parent_pid = os.getppid()
    os.kill(parent_pid, 9)
```

In [185]:
!http :3001/kill

HTTP/1.1 200 OK
content-length: 4
content-type: application/json
date: Fri, 24 Mar 2023 20:15:38 GMT
server: uvicorn

null




In [186]:
!docker-compose --project-directory visits ps

NAME                    COMMAND                  SERVICE             STATUS              PORTS
visits-redis-server-1   "docker-entrypoint.s…"   redis-server        running             6379/tcp
visits-web-server-1     "/bin/sh -c 'gunicor…"   web-server          restarting          


Note restarting status. After the service starts, the previous number of visits is persisted:

In [187]:
!http :3001/

HTTP/1.1 200 OK
content-length: 19
content-type: text/html; charset=utf-8
date: Fri, 24 Mar 2023 20:15:42 GMT
server: uvicorn

Number of visits: 3




<br>

```{figure} diagrams/01-restart-failure.png
---
width: 650px
name: 01-restart-failure
---
Logs when the web service exits with code 173. Notice it immediately restarts.
```

Stopping does not start the service. For this we need the other restart policies.

In [188]:
!docker-compose --project-directory visits ps -q web-server

4fdf0034399b797818684e15acfa93ac87205a348d4a22cbd887d568e66ef9d8


In [189]:
!docker stop 4fdf0034399b7

4fdf0034399b7


In [190]:
!docker-compose --project-directory visits ps

NAME                    COMMAND                  SERVICE             STATUS              PORTS
visits-redis-server-1   "docker-entrypoint.s…"   redis-server        running             6379/tcp
visits-web-server-1     "/bin/sh -c 'gunicor…"   web-server          exited (137)        


### Compose down

Stopping and teardown of all the services:

In [191]:
!docker-compose --project-directory visits down

[+] Running 1/0
 ⠿ Container visits-web-server-1    Rem...                                 0.0s
 ⠋ Container visits-redis-server-1  S...                                   0.1s
[+] Running 1/2
 ⠿ Container visits-web-server-1    Rem...                                 0.0s
 ⠙ Container visits-redis-server-1  S...                                   0.2s
[+] Running 2/2
 ⠿ Container visits-web-server-1    Rem...                                 0.0s
 ⠿ Container visits-redis-server-1  R...                                   0.2s
 ⠋ Network visits_default           Removing                               0.0s
[+] Running 3/3
 ⠿ Container visits-web-server-1    Rem...                                 0.0s
 ⠿ Container visits-redis-server-1  R...                                   0.2s
 ⠿ Network visits_default           Removed                                0.1s
